### DISCUSSION TITLE
8 bit image from 8 bit image...

### DISCUSSION THREAD

**Discussion by afwn90cj93201nixr2e1re** [ImageMagick/ImageMagick#1961] [General] [answered]
./del

**Comment by afwn90cj93201nixr2e1re** [comment, marked as answer]
@fmw42

**Reply by fmw42** [reply]
`convert blue.png 123.bmp -alpha off -compose over -composite -type palette result.bmp`

But, I do not think you can preserve the colormap in ImageMagick. It shuffles it around and there is no way that I know to set it the way you want, if it does not come preserved.

**Reply by fmw42** [reply]
Sorry, I do not know IMagick well. But Imagick uses ImageMagick to do the work and ImageMagick cannot manipulate nor necessarily preserve the color map. The output can be made as Palette 8-bit colormap, but cannot be necessarily saved with the same values or order as the input. My command above does make a palette output. I used blue.png since that was the image you posted. But you can create a blue image and do the same. In your Imagick code, you need to set the type (and perhaps not the quantization). If the blue is already in your color map, then quantization is not needed. Quantization is probably not needed as ImageMagick will quantize to max 256 colors if you set the type to palette.  See https://www.php.net/manual/en/imagick.settype.php

**Reply by fmw42** [reply]
Did you add setType() and remove quantize? What was your setType() command that you added?

**Reply by fmw42** [reply]
But you can setType in PHP Imagick from the link I sent above.

**Reply by fmw42** [reply]
Are you running ImageMagick from the install directory? That is a bad idea if you are. Create a new directory and put your images there and process it there. Do not use ./magick

`result.bmp BMP 512x512 512x512+0+0 8-bit sRGB 256c 8004B 0.000u 0:00.010`

It says 256c. That means 8-bit color not 24-bit color. So the result is palette and not tricolor. So the command is working fine.

**Reply by fmw42** [reply]
Imagick::setType ( int $image_type )

See PreDefined Constants in the PHP Imagick documentation and use:

imagick::IMGTYPE_PALETTE

**Reply by fmw42** [reply]
I am not a Windows user, so I do not know that you need -compress none for bmp file. Add it if it helps.

**Reply by fmw42** [reply]
Put setType just before writing the output, not before you create your composite.

**Reply by fmw42** [reply]
I am not an IMagick user, so I do not know. In ImageMagick, you read the input, add settings that control operators, then put the operators, then put settings that control the output, then write the output. They need to be done in the proper order (esp. for IM 7). Add compress none equivalent if you think that was important.

**Reply by fmw42** [reply]
try

Note the BMP: in the write statement


```
	foreach ($dir as $fileinfo) {
		if ($fileinfo->isDir())continue;
		$path = $fileinfo->getPathname();
		$im = new imagick($path);
		$output = new Imagick();
		$background = new \ImagickPixel('rgb(0, 0, 255)');
		$output->newimage(512,512,$background);
		$output->setImageCompression(Imagick::COMPRESSION_NO);
		$output->compositeimage($im->getimage(), Imagick::COMPOSITE_COPY, 0, 0);
		/* $output->quantizeImage(256, Imagick::COLORSPACE_SRGB, false, false, false); */
		$outpath = str_replace('bmps', 'bmps2', $path);
		$output->setImageFormat("bmp");
		$output->setImageDepth(8);
		$output->setType(Imagick::IMGTYPE_PALETTE);
		$output->writeImage(BMP:$outpath);
		$output->destroy();
		$im->destroy();
	}

```


If that fails, then use

`		$output->writeImage(BMP3:$outpath);
`

or

`		$output->writeImage(BMP2:$outpath);
`

**Reply by fmw42** [reply]
How do you know it does not work?  Do

identify on your output result in command line

**Reply by fmw42** [reply]
Move

`$output->setImageCompression(Imagick::COMPRESSION_NO);`

to after the composite. As I said before, output control must come after your processing.

**Reply by fmw42** [reply]
I do not understand the question. You can apply a palette to a grayscale image using an image of the colors you want to use in the palette and -clut to colorize it with those colors. See https://imagemagick.org/Usage/color_mods/#clut

Is that what you want?

**Reply by fmw42** [reply]
Well, then please explain in more detail what you want.

**Reply by snibgo** [reply]
> this image containt all of bitmap colors, so, can i generate same thing via imagick? For example just palette in bmp, in one line, just 256 colors 1x256 for example.

You can create a Nx1 image containing all the N colours of an input like this:
`magick in.png -unique-colors out.png`

But there is no guarantee the palette will be the same. It might be if you use "-define png:preserve-colormap=true".
